internal/clients/aws.go

```diff
@@ -7,61 +7,165 @@ package clients
+	keyAccessKeyID               = "access_key"
+	keySecretAccessKey           = "secret_key"
+	keyAssumeRoleWithWebIdentity = "assume_role_with_web_identity"
+	keyWebIdentityTokenFile      = "web_identity_token_file"
+	keyAssumeRole                = "assume_role"
+	keyTags                      = "tags"
+	keyTransitiveTagKeys         = "transitive_tag_keys"
 )
 
-// TerraformSetupBuilder returns Terraform setup with provider specific
-// configuration like provider credentials used to connect to cloud APIs in the
-// expected form of a Terraform provider.
-func TerraformSetupBuilder(version, providerSource, providerVersion string) terraform.SetupFn { //nolint:gocyclo
-	return func(ctx context.Context, client client.Client, mg resource.Managed) (terraform.Setup, error) {
-		cfg, err := GetAWSConfig(ctx, client, mg)
+func SelectTerraformSetup(version, providerSource, providerVersion string) terraform.SetupFn {
+	return func(ctx context.Context, c client.Client, mg resource.Managed) (terraform.Setup, error) {
+		pc := &v1beta1.ProviderConfig{}
+		if err := c.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
+			return terraform.Setup{}, errors.Wrap(err, "cannot get referenced Provider")
+		}
+		ps := terraform.Setup{
+			Version: version,
+			Requirement: terraform.ProviderRequirement{
+				Source:  providerSource,
+				Version: providerVersion,
+			},
+		}
+		account, err := getAccountId(ctx, c, mg)
+		if err != nil {
+			return terraform.Setup{}, errors.Wrap(err, "cannot get account id")
+		}
+		ps.ClientMetadata = map[string]string{
+			keyAccountId: account,
+		}
+
+		var err2 error
+		if len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil {
+			ps, err2 = DefaultTerraformSetupBuilder(ctx, c, mg, ps)
+			if err2 != nil {
+				return terraform.Setup{}, errors.Wrap(err2, "cannot build terraform configuration")
+			}
+		} else {
+			ps, err2 = OptimizedTerraformSetupBuilder(ctx, c, mg, pc, ps)
+			if err2 != nil {
+				return terraform.Setup{}, errors.Wrap(err2, "cannot build terraform configuration")
+			}
+		}
+		return ps, nil
+	}
+}
+
+func OptimizedTerraformSetupBuilder(ctx context.Context, c client.Client, mg resource.Managed, pc *v1beta1.ProviderConfig, ps terraform.Setup) (terraform.Setup, error) { //nolint:gocyclo
+	cfg, err := getAWSConfig(ctx, c, mg)
+	if err != nil {
+		return terraform.Setup{}, errors.Wrap(err, "cannot get AWS config")
+	}
+	ps.Configuration = map[string]any{
+		keyRegion: cfg.Region,
+	}
+
+	switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
+	case authKeyWebIdentity:
+		if pc.Spec.Credentials.WebIdentity == nil {
+			return terraform.Setup{}, errors.New(`spec.credentials.webIdentity of ProviderConfig cannot be nil when the credential source is "WebIdentity"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = pc.Spec.Credentials.WebIdentity.RoleARN
+		ps.Configuration[keyWebIdentityTokenFile] = defaultIdentityTokenFile
```

Comment:
We had better use the token file specified in the environment:
```suggestion
		ps.Configuration[keyWebIdentityTokenFile] = os.Getenv(envWebIdentityTokenFile)
```